components/public-api-server/pkg/proxy/conn.go

```diff
@@ -42,3 +46,69 @@ func (p *NoConnectionPool) Get(ctx context.Context, token string) (gitpod.APIInt
 
 	return server, nil
 }
+
+func NewConnectionPool(address *url.URL, poolSize int) (*ConnectionPool, error) {
+	cache, err := lru.NewWithEvict(poolSize, func(key, value interface{}) {
+		connectionPoolSize.Dec()
+
+		// We attempt to gracefully close the connection
+		conn, ok := value.(gitpod.APIInterface)
+		if !ok {
+			return
+		}
+
+		closeErr := conn.Close()
+		if closeErr != nil {
+			log.Log.WithError(closeErr).Warn("Failed to close connection to server.")
+		}
+	})
+	if err != nil {
+		return nil, fmt.Errorf("faield to create LRU cache: %w", err)
+	}
+
+	return &ConnectionPool{
+		cache: cache,
+		connConstructor: func(token string) (gitpod.APIInterface, error) {
+			return gitpod.ConnectToServer(address.String(), gitpod.ConnectToServerOpts{
+				// We want the connection to persist beyond the lifecycle of a single request
+				Context: context.Background(),
+				Token:   token,
+				Log:     log.Log,
+				CloseHandler: func(err error) {
+					cache.Remove(token)
+					connectionPoolSize.Dec()
+				},
+			})
+		},
+	}, nil
+
+}
+
+type ConnectionPool struct {
+	connConstructor func(token string) (gitpod.APIInterface, error)
+
+	// cache stores token to connection mapping
+	cache *lru.Cache
+}
+
+func (p *ConnectionPool) Get(ctx context.Context, token string) (gitpod.APIInterface, error) {
+	cached, found := p.cache.Get(token)
+	reportCacheOutcome(found)
+	if found {
+		connectionPoolCacheOutcome.WithLabelValues("true").Inc()
```

Comment:
I think this will increment the metric twice, given the call to `reportCacheOutcome` 2 lines above.

Reply:
Good spot. I refactored the metric but forgot to remove it here.